zetaclient/chains/evm/signer/outbound_data.go

```diff
@@ -45,143 +46,159 @@ type OutboundData struct {
+	switch {
+	case err != nil:
+		return nil, true, errors.Wrap(err, "failed to check if outbound is processed")
+	case alreadyIncluded || alreadyConfirmed:
+		logger.Info().Msgf("CCTX already processed; skipping")
+		return nil, true, nil
 	}
-	return nil
-}
 
-// NewOutboundData populates transaction input fields parsed from the cctx and other parameters
-// returns
-//  1. New NewOutboundData Data struct or nil if an error occurred.
-//  2. bool (skipTx) - if the transaction doesn't qualify to be processed the function will return true, meaning that this
-//     cctx will be skipped and false otherwise.
-//  3. error
-func NewOutboundData(
-	cctx *types.CrossChainTx,
-	evmObserver *observer.Observer,
-	evmRPC interfaces.EVMRPCClient,
-	logger zerolog.Logger,
-	height uint64,
-) (*OutboundData, bool, error) {
-	txData := OutboundData{}
-	txData.outboundParams = cctx.GetCurrentOutboundParam()
-	txData.amount = cctx.GetCurrentOutboundParam().Amount.BigInt()
-	txData.nonce = cctx.GetCurrentOutboundParam().TssNonce
-	txData.sender = ethcommon.HexToAddress(cctx.InboundParams.Sender)
-	txData.srcChainID = big.NewInt(cctx.InboundParams.SenderChainId)
-	txData.asset = ethcommon.HexToAddress(cctx.InboundParams.Asset)
-
-	txData.height = height
-
-	skipTx := txData.SetChainAndSender(cctx, logger)
-	if skipTx {
+	// Determine destination chain and address
+	to, toChainID, skip := determineDestination(cctx, logger)
+	if skip {
 		return nil, true, nil
 	}
 
-	toChain := chains.GetChainFromChainID(txData.toChainID.Int64())
+	toChain := chains.GetChainFromChainID(toChainID.Int64())
 	if toChain == nil {
-		return nil, true, fmt.Errorf("unknown chain: %d", txData.toChainID.Int64())
+		return nil, false, fmt.Errorf("unknown chain: %d", toChainID.Int64())
 	}
 
-	// Get nonce, Early return if the cctx is already processed
-	nonce := cctx.GetCurrentOutboundParam().TssNonce
-	included, confirmed, err := evmObserver.IsOutboundProcessed(cctx, logger)
-	if err != nil {
-		return nil, true, errors.New("IsOutboundProcessed failed")
-	}
-	if included || confirmed {
-		logger.Info().Msgf("CCTX already processed; exit signer")
-		return nil, true, nil
-	}
+	nonce := outboundParams.TssNonce
 
-	// Set up gas limit and gas price
-	err = txData.SetupGas(cctx, logger, evmRPC, toChain)
+	// Get sendHash
+	cctxIndex, err := getCCTXIndex(cctx)
 	if err != nil {
-		return nil, true, err
+		return nil, false, err
 	}
 
-	// Get sendHash
-	logger.Info().
-		Msgf("chain %s minting %d to %s, nonce %d, finalized zeta bn %d", toChain, cctx.InboundParams.Amount, txData.to.Hex(), nonce, cctx.InboundParams.FinalizedZetaHeight)
-	cctxIndex, err := hex.DecodeString(cctx.Index[2:]) // remove the leading 0x
-	if err != nil || len(cctxIndex) != 32 {
-		return nil, true, fmt.Errorf("decode CCTX %s error", cctx.Index)
+	// Determine gas fees
+	gas, err := makeGasFromCCTX(cctx, evmObserver.GetPriorityGasFee(), logger)
```

Comment:
If the `priorityFee ` being used here comes from zetaclients own fetched value thru RPC `SuggestGasTipCap`.  The values may vary across across the group of TSS signers (the external signers are using their own Ethereum node/endpoint, and these nodes can go slightly out-of-sync). In this case, the tx body being built across TSS signers will also vary and lead to TSS keysign failure (because they're signing different transaction bodies).

The key point to achieve EIP-1559 is to be able to use a deterministic `GasTipCap` value across all TSS signers to sign the transaction. So the cctx struct itself will need to carry a consensus-based `GasTipCap`, with zetaclients just use the same value to create the tx and then process TSS keysign.

Reply:
Good point on this. We will have to post some information on ZetaCore finally.
We can add a new field in the `MsgVoteGasPrice` and in the `GasPrice` structure in the `crosschain` module.
(We could think about renaming the `GasPrice` structure in the future but right now we could keep the name and just add new fields)

Unlike the actual gas price, here we would not need to interact with the smart contract like `fungibleKeeper.SetGasPrice`. We would just need to update the structure at the Cosmos module level, it will then be read by all ZetaClient.

I think we can keep this PR open, work on a new one for the gas tip posting, then update this one.

WDYT @swift1337

Reply:
Makes total sense for me, thanks for spotting this, @ws4charlie.